src/traces/scattermapbox/plot.js

```diff
@@ -85,51 +86,96 @@ proto.update = function update(calcTrace) {
     var optsAll = convert(subplot.gd, calcTrace);
     var below = subplot.belowLookup['trace-' + this.uid];
-    var i, k, opts;
     var hasCluster = !!(trace.cluster && trace.cluster.enabled);
     var hadCluster = !!this.clusterEnabled;
-
-    if(below !== this.below) {
-        var order = ORDER.nonCluster;
-
-        for(i = order.length - 1; i >= 0; i--) {
-            k = order[i];
-            map.removeLayer(this.layerIds[k]);
+    var lThis = this;
+
+    function addCluster(noSource) {
+        if(!noSource) lThis.addSource('circle', optsAll.circle, trace.cluster);
+        var order = ORDER.cluster;
+        for(var i = 0; i < order.length; i++) {
+            var k = order[i];
+            var opts = optsAll[k];
+            lThis.addLayer(k, opts, below);
         }
-        for(i = 0; i < order.length; i++) {
-            k = order[i];
-            opts = optsAll[k];
-            this.addLayer(k, opts, below);
-        }
-        this.below = below;
-    } else if(hasCluster && !hadCluster) {
-        for(i = ORDER.nonCluster.length - 1; i >= 0; i--) {
-            k = ORDER.nonCluster[i];
-            map.removeLayer(this.layerIds[k]);
-            map.removeSource(this.sourceIds[k]);
+    }
+
+    function removeCluster(noSource) {
+        var order = ORDER.cluster;
+        for(var i = order.length - 1; i >= 0; i--) {
+            var k = order[i];
+            map.removeLayer(lThis.layerIds[k]);
         }
-        this.addSource('circle', optsAll.circle, trace.cluster);
-        for(i = 0; i < ORDER.cluster.length; i++) {
-            k = ORDER.cluster[i];
-            opts = optsAll[k];
-            this.addLayer(k, opts, below);
+        if(!noSource) map.removeSource(lThis.sourceIds.circle);
+    }
+
+    function addNonCluster(noSource) {
+        var order = ORDER.nonCluster;
+        for(var i = 0; i < order.length; i++) {
+            var k = order[i];
+            var opts = optsAll[k];
+            if(!noSource) lThis.addSource(k, opts);
+            lThis.addLayer(k, opts, below);
         }
-        this.clusterEnabled = hasCluster;
-    } else if(!hasCluster && hadCluster) {
-        for(i = 0; i < ORDER.cluster.length; i++) {
-            k = ORDER.cluster[i];
-            map.removeLayer(this.layerIds[k]);
+    }
+
+    function removeNonCluster(noSource) {
+        var order = ORDER.nonCluster;
+        for(var i = order.length - 1; i >= 0; i--) {
+            var k = order[i];
+            map.removeLayer(lThis.layerIds[k]);
+            if(!noSource) map.removeSource(lThis.sourceIds[k]);
         }
-        map.removeSource(this.sourceIds.circle);
-        for(i = 0; i < ORDER.nonCluster.length; i++) {
-            k = ORDER.nonCluster[i];
-            opts = optsAll[k];
-            this.addSource(k, opts, trace.cluster);
-            this.addLayer(k, opts, below);
+    }
+
+    function remove(noSource) {
+        if(hadCluster) removeCluster(noSource); else removeNonCluster(noSource);
```

Comment:
not a big deal, but if you really want this to be a one-liner you can collapse it more; otherwise I'd write it as multiple lines. (same in `add` below)
```suggestion
        (hadCluster ? removeCluster : removeNonCluster)(noSource);
```